llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -142,9 +152,124 @@ PreservedAnalyses CompileTimePropertiesPass::run(Module &M,
     }
   }
 
+  // Check pointer annotations.
+  SmallVector<IntrinsicInst *, 4> RemovableAnnots;
+  for (Function &F : M)
+    for (BasicBlock &BB : F)
+      for (Instruction &Inst : BB)
+        if (auto *IntrInst = dyn_cast<IntrinsicInst>(&Inst))
+          if (IntrInst->getIntrinsicID() == Intrinsic::ptr_annotation &&
+              transformSYCLPropertiesAnnotation(M, IntrInst, RemovableAnnots))
+            CompileTimePropertiesMet = true;
+
+  // Remove irrelevant "sycl-properties" annotations after the transformations.
+  for (IntrinsicInst *IntrInst : RemovableAnnots) {
+    assert(IntrInst->getNumUses() == 0);
+    IntrInst->eraseFromParent();
+  }
+
   // The pass just adds some metadata to the module, it should not ruin
   // any analysis, but we need return PreservedAnalyses::none() to inform
   // the caller that at least one compile-time property was met.
   return CompileTimePropertiesMet ? PreservedAnalyses::none()
                                   : PreservedAnalyses::all();
 }
+
+bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
+    Module &M, IntrinsicInst *IntrInst,
+    SmallVector<IntrinsicInst *, 4> &RemovableAnnotations) {
+  assert(IntrInst->getIntrinsicID() == Intrinsic::ptr_annotation &&
+         "Intrinsic is not a pointer annotation.");
+
+  // Get the global variable with the annotation string.
+  const GlobalVariable *AnnotStrArgGV = nullptr;
+  if (auto *GEP = dyn_cast<GEPOperator>(IntrInst->getArgOperand(1)))
+    if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+      AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  if (!AnnotStrArgGV)
+    return false;
+
+  // We only need to consider annotations with "sycl-properties" annotation
+  // string.
+  Optional<StringRef> AnnotStr = getGlobalVariableString(AnnotStrArgGV);
+  if (!AnnotStr || AnnotStr->str() != "sycl-properties")
+    return false;
+
+  // Read the annotation values and create the new annotation string.
+  std::string NewAnnotString = "";
+  if (const auto *Cast =
+          dyn_cast<BitCastOperator>(IntrInst->getArgOperand(4))) {
```

Comment:
What about to add an assert to check that the `IntrInst` has 5 operands?

Reply:
The intrinsic should always have exactly 5 operands, based on its definition. To be safe though, I have added an assert at the beginning of the transformation function.